boards/opentitan/src/tests/flash.rs

```diff
@@ -2,417 +2,474 @@
-    fn read_complete(&self, page: &'static mut F::Page, result: Result<(), hil::flash::Error>) {
-        if self.read_pending.get() {
-            if result.is_err() {
-                let error = result.unwrap_err();
-                assert_eq!(error, hil::flash::Error::FlashMemoryProtectionError);
-                self.mp_fault_detect.set(true);
+mod protections_and_controller {
+    use crate::tests::run_kernel_op;
+    use crate::PERIPHERALS;
+    use core::cell::Cell;
+    use kernel::debug;
+    use kernel::errorcode::ErrorCode;
+    use kernel::hil;
+    use kernel::hil::flash::Flash;
+    use kernel::hil::flash::HasClient;
+    use kernel::static_init;
+    use kernel::utilities::cells::TakeCell;
+    use lowrisc::flash_ctrl::FlashMPConfig;
+    use lowrisc::flash_ctrl::FLASH_ADDR_OFFSET;
+    use lowrisc::flash_ctrl::PAGE_SIZE;
+
+    struct FlashCtlCallBack {
+        read_pending: Cell<bool>,
+        write_pending: Cell<bool>,
+        // A lowrisc page to for reads/writes
+        read_in_page: TakeCell<'static, lowrisc::flash_ctrl::LowRiscPage>,
+        write_in_page: TakeCell<'static, lowrisc::flash_ctrl::LowRiscPage>,
+        // We recover the callback returned buffer into these
+        read_out_buf: TakeCell<'static, [u8]>,
+        write_out_buf: TakeCell<'static, [u8]>,
+        // Flag if an MP fault was detected
+        mp_fault_detect: Cell<bool>,
+    }
+
+    impl<'a> FlashCtlCallBack {
+        fn new(
+            read_in_page: &'static mut lowrisc::flash_ctrl::LowRiscPage,
+            write_in_page: &'static mut lowrisc::flash_ctrl::LowRiscPage,
+        ) -> Self {
+            FlashCtlCallBack {
+                read_pending: Cell::new(false),
+                write_pending: Cell::new(false),
+                read_in_page: TakeCell::new(read_in_page),
+                write_in_page: TakeCell::new(write_in_page),
+                read_out_buf: TakeCell::empty(),
+                write_out_buf: TakeCell::empty(),
+                mp_fault_detect: Cell::new(false),
             }
-            self.read_out_buf.replace(page.as_mut());
+        }
+
+        fn reset(&self) {
             self.read_pending.set(false);
+            self.write_pending.set(false);
+            self.mp_fault_detect.set(false);
         }
     }
 
-    fn write_complete(&self, page: &'static mut F::Page, result: Result<(), hil::flash::Error>) {
-        if self.write_pending.get() {
-            if result.is_err() {
-                let error = result.unwrap_err();
-                assert_eq!(error, hil::flash::Error::FlashMemoryProtectionError);
-                self.mp_fault_detect.set(true);
+    impl<'a, F: hil::flash::Flash> hil::flash::Client<F> for FlashCtlCallBack {
+        fn read_complete(&self, page: &'static mut F::Page, error: Result<(), hil::flash::Error>) {
+            if self.read_pending.get() {
+                if error == Err(hil::flash::Error::FlashMemoryProtectionError) {
```

Comment:
```suggestion
                if let Err(hil::flash::Error::FlashMemoryProtectionError) = err {
```

Reply:
Is this an improvement? We aren't using the error variable so I don't see this being any cleaner